### PR TITLE
Do not set LSN for new FPI page

### PR DESCRIPTION
**Pull request by knizhnik**
refer #1656

### CODE REVIEW

**Review by ololobus — COMMENTED** [review]
Postgres (compute) now panics with `PANIC:  Page 0 of relation 1663/16384/21210.0 is evicted with zero LSN` on our LSN check in zenith smgr.

That's a bit weird. OK, we got a new page with WAL record and stored it in the pageserver without stamping the LSN. Then Postgres got this page from PS (this should work the same way as vanilla Postgres reads new pages after smgrextend, for example, IIUC). But why it tries to evict non-new page with zero LSN?

File: pageserver/src/walingest.rs [hunk 1]
```diff
@@ -304,8 +304,11 @@ impl<'a, R: Repository> WalIngest<'a, R> {
                 image.resize(image.len() + blk.hole_length as usize, 0u8);
                 image.unsplit(tail);
             }
-            image[0..4].copy_from_slice(&((lsn.0 >> 32) as u32).to_le_bytes());
-            image[4..8].copy_from_slice(&(lsn.0 as u32).to_le_bytes());
+            if image[10] != 0 || image[11] != 0 {
```
Comment (hlinnaka) [inline pageserver/src/walingest.rs:307]: Hmm, this assumes that the page uses the "standard page layout", i.e PageHeaderData. That's not guaranteed. However, I see that XLogReadBufferForRedoExtended() makes the same assumption. It doesn't seem safe in upstream either...

Comment (hlinnaka) [reply pageserver/src/walingest.rs:307]: Let's add a PageIsNew() function, to make it clear what this does, at least, and add a comment explaining that this matches the behaviour of XLogReadBufferForRedoExtended()

**Comment by knizhnik** [comment]
> Postgres (compute) now panics with `PANIC: Page 0 of relation 1663/16384/21210.0 is evicted with zero LSN` on our LSN check in zenith smgr.
> 
> That's a bit weird. OK, we got a new page with WAL record and stored it in the pageserver without stamping the LSN. Then Postgres got this page from PS (this should work the same way as vanilla Postgres reads new pages after smgrextend, for example, IIUC). But why it tries to evict non-new page with zero LSN?

For example spgist can produce dirty page with zero LSN:
```
	SpGistInitMetapage(BufferGetPage(metabuffer));
	MarkBufferDirty(metabuffer);
	SpGistInitBuffer(rootbuffer, SPGIST_LEAF);
	MarkBufferDirty(rootbuffer);
	SpGistInitBuffer(nullbuffer, SPGIST_LEAF | SPGIST_NULLS);
	MarkBufferDirty(nullbuffer);

	END_CRIT_SECTION();

	UnlockReleaseBuffer(metabuffer);
	UnlockReleaseBuffer(rootbuffer);
	UnlockReleaseBuffer(nullbuffer);
```

I wonder if just replacing ereport(PANIC,...) with ereport(SmgrTrace, ...) is acceptable solution?
At lease pg_regress test is passed after such change.

**Comment by hlinnaka** [comment]
> > Postgres (compute) now panics with `PANIC: Page 0 of relation 1663/16384/21210.0 is evicted with zero LSN` on our LSN check in zenith smgr.
> > That's a bit weird. OK, we got a new page with WAL record and stored it in the pageserver without stamping the LSN. Then Postgres got this page from PS (this should work the same way as vanilla Postgres reads new pages after smgrextend, for example, IIUC). But why it tries to evict non-new page with zero LSN?
> 
> For example spgist can produce dirty page with zero LSN

That particular example is handled by the special "unlogged index build" mechanism, see thethe `smgr_start_unlogged_build(index->rd_smgr)` call before that piece of code.

> I wonder if just replacing ereport(PANIC,...) with ereport(SmgrTrace, ...) is acceptable solution? At lease pg_regress test is passed after such change.

No, that is an important safeguard. Firstly, if a page is created without WAL-logging, that's a problem with Neon, because if it's not WAL-logged it is lost. So if a page appears in the buffer cache that's not stamped by an LSN, that's a problem. Ideally, we would catch that case earlier - maybe we should add an assertion into ReleaseBuffer() - but silently losing the page is not OK.

Secondly, assuming the page was WAL-logged but we just didn't set the LSN in the cached version of the page, we need to know the LSN of the evicted page, so that we know to use that LSN in subsequent GetPage requests. Otherwise we might fail to read back the page, even if it was stored in the pageserver correctly. As a workaround for that, we could bump up the last-evicted LSN to the last-flushed LSN when a page with zero LSN is evicted. But we cannot distinguish that from the first case.

File: pageserver/src/walingest.rs (hunk 1 above)
Comment (hlinnaka) [reply pageserver/src/walingest.rs]: Okay, good point on the checksums..

**Comment by knizhnik** [comment]
> So if a page appears in the buffer cache that's not stamped by an LSN, that's a problem. Ideally, we would catch that case earlier - maybe we should add an assertion into ReleaseBuffer() - but silently losing the page is not OK.

Actually, I have already done it: this is how I produced this stack trace above.
But you are right - it is special case: building gin/gist/spgist indexes.
Ignoring this case I found the only place where it is possible that buffer is dirties and released without setting lsn: it is spgvacuumscan (during pg_regssion test). spgistvacuum calls SpGistUpdateMetaPage which does the following:
```
		if (ConditionalLockBuffer(metabuffer))
		{
			Page		metapage = BufferGetPage(metabuffer);
			SpGistMetaPageData *metadata = SpGistPageGetMeta(metapage);

			metadata->lastUsedPages = cache->lastUsedPages;

			/*
			 * Set pd_lower just past the end of the metadata.  This is
			 * essential, because without doing so, metadata will be lost if
			 * xlog.c compresses the page.  (We must do this here because
			 * pre-v11 versions of PG did not set the metapage's pd_lower
			 * correctly, so a pg_upgraded index might contain the wrong
			 * value.)
			 */
			((PageHeader) metapage)->pd_lower =
				((char *) metadata + sizeof(SpGistMetaPageData)) - (char *) metapage;
			MarkBufferDirty(metabuffer);
			UnlockReleaseBuffer(metabuffer);
		}
```
If I add here:
```
			PageSetLSN(metapage, GetXLogInsertRecPtr());
```
then there is no panic any more.
I still not sure whether it means problem in mainstream postgres and what is the proper fix in this case.

**Comment by hlinnaka** [comment]
> Actually, I have already done it: this is how I produced this stack trace above. But you are right - it is special case: building gin/gist/spgist indexes. Ignoring this case I found the only place where it is possible that buffer is dirties and released without setting lsn: it is spgvacuumscan (during pg_regssion test). spgistvacuum calls SpGistUpdateMetaPage which does the following:
> 
> ```
> 		if (ConditionalLockBuffer(metabuffer))
> 		{
> 			Page		metapage = BufferGetPage(metabuffer);
> 			SpGistMetaPageData *metadata = SpGistPageGetMeta(metapage);
> 
> 			metadata->lastUsedPages = cache->lastUsedPages;
> 
> 			/*
> 			 * Set pd_lower just past the end of the metadata.  This is
> 			 * essential, because without doing so, metadata will be lost if
> 			 * xlog.c compresses the page.  (We must do this here because
> 			 * pre-v11 versions of PG did not set the metapage's pd_lower
> 			 * correctly, so a pg_upgraded index might contain the wrong
> 			 * value.)
> 			 */
> 			((PageHeader) metapage)->pd_lower =
> 				((char *) metadata + sizeof(SpGistMetaPageData)) - (char *) metapage;
> 			MarkBufferDirty(metabuffer);
> 			UnlockReleaseBuffer(metabuffer);
> 		}
> ```
> 
> If I add here:
> 
> ```
> 			PageSetLSN(metapage, GetXLogInsertRecPtr());
> ```
> 
> then there is no panic any more. I still not sure whether it means problem in mainstream postgres and what is the proper fix in this case.

Hmm, so the metadata-page update isn't WAL-logged, on purpose. Seems fine in PostgreSQL, but for Neon it means that the lastUsedPages cache is lost when the page is evicted from the buffer cache. That's probably fine, too, except for the PANIC it induces.

Why did the metadata page have zero LSN in the first place, though? Surely the change that created the page was WAL-logged.

**Comment by knizhnik** [comment]
Sorry, I did stupid mistake in calculation of offsets in PagHeader struct.
Now there are no panics any more and not changes in vendor/postgres are needed.

File: pageserver/src/walingest.rs [hunk 2]
```diff
@@ -304,8 +305,9 @@ impl<'a, R: Repository> WalIngest<'a, R> {
                 image.resize(image.len() + blk.hole_length as usize, 0u8);
                 image.unsplit(tail);
             }
-            image[0..4].copy_from_slice(&((lsn.0 >> 32) as u32).to_le_bytes());
-            image[4..8].copy_from_slice(&(lsn.0 as u32).to_le_bytes());
+            if !page_is_new(&image) {
```
Comment (hlinnaka) [inline pageserver/src/walingest.rs:308]: Please add a comment that this needs to match the logic in `XLogReadBufferForRedoExtended`

**Review by hlinnaka — APPROVED** [review]

File: pageserver/src/walingest.rs (hunk 2 above)
Comment (knizhnik) [reply pageserver/src/walingest.rs:313]: Done

**Pull request merged**